File: Algorithms/Easy/LongestCommonPrefix.go
package easy

// Time complexity: O(n)
// Space complexity: O(1)
func longestCommonPrefix(strs []string) string {
	res := make([]byte, 0)
	for i := range strs[0] {
		for _, str := range strs {
			if i == len(str) || str[i] != strs[0][i] {
				return string(res)
			}
		}
		res = append(res, strs[0][i])
	}

	return string(res)
}

File: Algorithms/Easy/14.Longest-Common-Prefix/LongestCommonPrefix.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def longestCommonPrefix(self, strs: list[str]) -> str:
        smallest = min(strs, key=len)
        for i, c in enumerate(smallest):
            for s in strs:
                if s[i] != c:
                    return smallest[:i]

        return smallest
